### PR TITLE
Release debug and interface adjustments

### PR DESCRIPTION
**Pull request by aaron0browne**
Not at all complete, but wanted to get it pushed so the review could begin.

### CODE REVIEW

File: errors.go [hunk 1]
```diff
@@ -93,11 +93,24 @@ type ValidationError struct {
...
-		return fmt.Sprintf("line %d: %s", e.Line, e.Err)
+		location = fmt.Sprintf("line %d", e.Line)
+	} else {
+		location = fmt.Sprintf("line %d, field %s", e.Line, e.Field)
+	}
+
+	if e.Context != nil {
+		context = fmt.Sprintf("%v", e.Context)
+	} else {
+		context = "no context provided"
```
Comment (bruth) [inline errors.go]: If a context isn't supplied, I presume there is none available. I am more in favor of it not providing a message if no context is available.

**Comment by aaron0browne** [comment]
Pretty happy with the bad header output now. One thing is that up until now we've been allowing sites to submit files that are entirely missing (non-required) columns that they don't submit data for. The validator will error on the header and every row of such a file. I think this calls for a convention change: sites should submit data with all columns, even if they leave an entire column NULL. This should align with automated DDL and export tools of the future.

File: errors.go [hunk 2]
```diff
@@ -93,11 +93,20 @@ type ValidationError struct {
...
-		return fmt.Sprintf("line %d: %s", e.Line, e.Err)
+		location = fmt.Sprintf("line %d", e.Line)
+	} else {
+		location = fmt.Sprintf("line %d, field %s", e.Line, e.Field)
+	}
+
+	if e.Context != nil {
+		return fmt.Sprintf("%s: %s: %v", location, e.Err, e.Context)
+	} else {
+		return fmt.Sprintf("%s: %s", location, e.Err)
```
Comment (bruth) [inline errors.go]: Is it preferred to not wrap the last `return` statement in an `else` block so it is clear what the last resort is.

File: validator.go [hunk 3]
```diff
@@ -82,29 +82,28 @@ func (t *TableValidator) validateRow(row []string) error {
 // a set of validators for each field.
 func (t *TableValidator) Init() error {
 	var (
-		err  error
-		head []string
+		err         error
+		head        []string
+		length_err  = false
+		match_err   = false
```
Comment (bruth) [inline validator.go]: Short variable names are idiomatic. If one word variable names are not reasonable, use camelcase.

Comment (bruth) [inline validator.go]: Also variables that are declared with the [zero value](https://golang.org/ref/spec#The_zero_value) can be just declared, `var matchErr bool` without the assignment.

File: validator.go [hunk 4]
```diff
@@ -129,13 +128,16 @@ func (t *TableValidator) Init() error {
 	// Set of fields by position mapped to schema.
 	t.fields = fields
 
-	if len(extra) > 0 || len(missing) > 0 {
+	err_context["unknown_fields"] = unknown
+	err_context["missing_fields"] = missing
+	if len(unknown) > 0 || len(missing) > 0 {
```
Comment (bruth) [inline validator.go]: Empty line above.

File: validator.go [hunk 5]
```diff
@@ -129,13 +128,16 @@ func (t *TableValidator) Init() error {
...
+
+	if length_err || match_err {
 		return &ValidationError{
-			Err: ErrExtraColumns,
-			Context: Context{
-				"extra":   extra,
-				"missing": missing,
-			},
+			Err:     ErrBadHeader,
+			Context: err_context,
```
Comment (bruth) [inline validator.go]: Since the context is going to have all four keys set, then there is no reason to incrementally assign the keys above. Just define the context here and set the fields like it was before.

**Comment by aaron0browne** [comment]
Updated based on your comments.

**Comment by aaron0browne** [comment]
It looks like the file to table matching does not do the fancy column set resolution promised in the usage message: https://github.com/chop-dbhi/data-models-validator/blob/master/cmd/validator/main.go#L158 I don't think we need to add the feature, I'm just planning to change the usage message, unless the feature exists somewhere and I'm missing it.

**Comment by bruth** [comment]
The filename should be set as the table [here](https://github.com/chop-dbhi/data-models-validator/blob/master/cmd/validator/main.go#L168). What do you mean by:

> fancy column set resolution

**Comment by aaron0browne** [comment]
I'm talking about this: https://github.com/chop-dbhi/data-models-validator/blob/master/cmd/validator/main.go#L36-L37

**Comment by bruth** [comment]
Ah right, no that is not implemented. You can remove it since it is not necessary.

**Comment by bruth** [comment]
This looks good in the current state. Any more work being done to the branch? If not, then merge.

**Comment by aaron0browne** [comment]
Still working on the error output. Would like some context in the samples, like what type is expected, for ErrTypeMismatch, etc. Also seeing some funkiness around unescaped quote errors not bubbling up.

**Comment by aaron0browne** [comment]
@bruth Can I get your help with the commit I just pushed? I started by trying to add context to the sample output in `main.go`, but then realized that was an `*Error` type not an `error` type... So then I changed some function signatures so an `error` would be returned, but now [this line](https://github.com/chop-dbhi/data-models-validator/commit/618c6d5939e5109dd95c1f309e54b3678fd5626d#diff-2c21fb78980d1744f15f8637f5c8700bL66) is resulting in a type error. I feel like I'm close, but what do I know?

**Comment by bruth** [comment]
> So then I changed some function signatures so an error would be returned, but now this line is resulting in a type error. I feel like I'm close, but what do I know?

The `ValidationError.Err` field is of type `*Error` which is one of the pre-defined errors. To remove the type error, you would need to change [this line](https://github.com/chop-dbhi/data-models-validator/blob/master/errors.go#L88) to `Err error`, but we want to keep it as type `*Error` so we can access the properties of the custom error type such as [here](https://github.com/chop-dbhi/data-models-validator/blob/master/cmd/validator/main.go#L252-L253).

**Comment by aaron0browne** [comment]
Some sites send us files with BOMs at the beginning.

I attempted to trim the BOM in the [UniversalReader.Read function](https://github.com/chop-dbhi/data-models-validator/commit/27711d6b3be1d9602e23c13faa91f2b8a8bc521e#diff-ed6ec976ee948cb81f2bbef47adb704dR33), but am still getting a header error. This seems to be because the TableValidator.csv.Read function is not inheriting the UniversalReader.Read method?

This last even though the [validator.Open function](https://github.com/chop-dbhi/data-models-validator/blob/27711d6b3be1d9602e23c13faa91f2b8a8bc521e/cmd/validator/main.go#L179) should return a UniversalReader-including reader, which is then used to instantiate the TableValidator in the [validator.New function](https://github.com/chop-dbhi/data-models-validator/blob/27711d6b3be1d9602e23c13faa91f2b8a8bc521e/cmd/validator/main.go#L186), which [creates the csvReader from that reader](https://github.com/chop-dbhi/data-models-validator/blob/27711d6b3be1d9602e23c13faa91f2b8a8bc521e/validator.go#L194).

Help would be appreciated. Apologies for any mis-use of terminology in describing the relations between bits of code.

I've left a [debugging statement](https://github.com/chop-dbhi/data-models-validator/commit/27711d6b3be1d9602e23c13faa91f2b8a8bc521e#diff-2c21fb78980d1744f15f8637f5c8700bR96) in the code and committed a test file so you can see the problem. After `make build`, do:

```
$ data-models-validator -model i2b2_pedsnet -version 2.0.0 BOM-test-validator-obs-fact.csv:observation_fact
Validating against model 'i2b2_pedsnet/2.0.0'
* Evaluating 'observation_fact' table in 'BOM-test-validator-obs-fact.csv'...
---
t.csv.Read
---
string: ["ENCOUNTER_NUM" PATIENT_NUM CONCEPT_CD PROVIDER_ID START_DATE MODIFIER_CD INSTANCE_NUM VALTYPE_CD TVAL_CHAR NVAL_NUM VALUEFLAG_CD QUANTITY_NUM UNITS_CD END_DATE LOCATION_CD OBSERVATION_BLOB CONFIDENCE_NUM UPDATE_DATE DOWNLOAD_DATE IMPORT_DATE SOURCESYSTEM_CD UPLOAD_ID]
hex: [efbbbf22454e434f554e5445525f4e554d22 50415449454e545f4e554d 434f4e434550545f4344 50524f56494445525f4944 53544152545f44415445 4d4f4449464945525f4344 494e5354414e43455f4e554d 56414c545950455f4344 5456414c5f43484152 4e56414c5f4e554d 56414c5545464c41475f4344 5155414e544954595f4e554d 554e4954535f4344 454e445f44415445 4c4f434154494f4e5f4344 4f42534552564154494f4e5f424c4f42 434f4e464944454e43455f4e554d 5550444154455f44415445 444f574e4c4f41445f44415445 494d504f52545f44415445 534f5552434553595354454d5f4344 55504c4f41445f4944]
---
* Problem reading CSV header: line 0: [code: 201] Header does not contain the correct set of fields: map[actual_length:22 unknown_fields:["encounter_num"] missing_fields:[encounter_num] expected_length:22]
```

That `efbbbf` bit at the beginning of the hex print out is the BOM character. The truly odd thing is that if you put a similar print statement in the UniversalReader.Read function, the BOM does not show up...

File: cmd/validator/main.go [hunk 6]
```diff
@@ -204,7 +204,7 @@ func main() {
 			"code",
 			"error",
 			"occurrences",
-			"sample",
+            "samples",
```
Comment (bruth) [inline cmd/validator/main.go]: Can you ensure all files have been run through `gofmt`?

File: validators.go [hunk 7]
```diff
@@ -50,7 +51,39 @@ var EncodingValidator = &Validator{
 				}
 			}
 
-			return ErrBadEncoding
+			return &ValidationError{
+				Err: ErrBadEncoding,
+				Context: Context{
+					"badRunes": bad,
```
Comment (bruth) [inline validators.go]: It appears most the context fields used elsewhere use underscores, not camel case. Since these are map keys I don't personally think they need to adhere to the camel case idiom, however they should be either or.

**Comment by aaron0browne** [comment]
Updates made in response to your comments. I'm going to test on more datasets today. Hopefully I won't find any more edge cases to resolve and I can move on to updating the README and usage.

**Comment by aaron0browne** [comment]
Let the README review begin! /cc @bruth @murphyke

File: README.md [hunk 8]
```diff
@@ -6,14 +6,140 @@ A validator for CSV files adhering to the [Data Models](https://github.com/chop-
...
+                        [-delim <delimiter>]
+                        [-compr <compression>]
+                        [-service <service>]
+                        ( <file>[:<table>]... | [:<table>] )
+
+The Data Models Validator reads a file containing data and checks it against
+the data model's schema. Input files or stream are delimited files (such as CSV)
+and optionally compressed using gzip or bzip2.
+
+One or more existing input files can be explicitly passed otherwise STDIN
```
Comment (murphyke) [inline README.md]: ... explicitly passed; otherwise, ...

File: README.md [hunk 9]
```diff
@@ -6,14 +6,140 @@ A validator for CSV files adhering to the [Data Models](https://github.com/chop-
...
+  data-models-validator -model omop -version 5.0.0 person.csv
+
+  # Validate foo.csv against the person table in the OMOP v5 data model.
+  data-models-validator -model omop -version 5.0.0 foo.csv:person
+
+  # Validate the STDIN stream denoting it is tab-delimited and gzipped.
+  data-models-validator -model omop -version 5.0.0 -delim $'\t' -compr gzip
+
+```
+
```
Comment (murphyke) [inline README.md]: Hopefully the validator returns an exit status of 0 for success and nonzero otherwise, and that should be documented.

Comment (bruth) [inline README.md]: @murphyke Good catch. It does everywhere except at the end! @aaron0browne Can you update [this block](https://github.com/chop-dbhi/data-models-validator/pull/7/files#diff-5208ca976625713cbeb5fa29135248beR262) to be:

```go
 		if nerrs > 0 {
 			fmt.Println("* A few issues were found")
 			tw.Render()
            os.Exit(1)
 		}
 		
        fmt.Println("* Everything looks good!")
```

File: README.md [hunk 10]
```diff
@@ -6,14 +6,140 @@ A validator for CSV files adhering to the [Data Models](https://github.com/chop-
...
+Download the latest binary from the [releases page](https://github.com/chop-dbhi/data-models-validator/releases) for your architecture: Windows, Linux, or OS X. The following examples assume the binary has been placed on your `PATH` with the name `data-models-validator`.
 
 ## Usage
 
 Run the following to see the usage.
 
 ```
-data-models-validator -help
+$ data-models-validator -help
+Data Models Validator - 0.1.0-beta.1
```
Comment (bruth) [inline README.md]: I am typically not a fan of embedding the usage in the README primarily because it will quickly get out of date, especially the version number. I understand that it is for example purposes, but it takes up a lot of space and could later cause confusion. What is correctly being shown is how to get the help/usage using the program.

File: README.md [hunk 11]
```diff
@@ -6,14 +6,140 @@ A validator for CSV files adhering to the [Data Models](https://github.com/chop-
...
+
+No harm is done to any data and by waiting several minutes before re-running, the problem can be circumvented. The development team is working on a fix ASAP and you can track our progress [here](https://github.com/chop-dbhi/data-models-validator/issues/9)
+
+## Future Directions
+
+Soon, we hope to add higher level validation checks, especially of referential integrity.
+
+## Output Examples
+
+If the filename does not match a table in the model, a list of known tables in the model is printed. The fix is to specify the table by adding `:table_name` to the end of your file name:
```
Comment (bruth) [inline README.md]: `The fix` &rarr; `The solution` sounds more appropriate since this was an intentional design.

File: README.md [hunk 12]
```diff
@@ -6,14 +6,140 @@ A validator for CSV files adhering to the [Data Models](https://github.com/chop-
...
+
+```
+$ data-models-validator -model pedsnet -version 2.0.0 PEDSNET_DRUG_EXPOSURE.csv
+Validating against model 'pedsnet/2.0.0'
+* Unknown table 'PEDSNET_DRUG_EXPOSURE'.
+Choices are: care_site, concept, concept_ancestor, concept_class, concept_relationship, concept_synonym, condition_occurrence, death, domain, drug_exposure, drug_strength, fact_relationship, location, measurement, observation, observation_period, person, procedure_occurrence, provider, relationship, source_to_concept_map, visit_occurrence, visit_payer, vocabulary
+$ data-models-validator -model pedsnet -version 2.0.0 PEDSNET_DRUG_EXPOSURE.csv:drug_exposure
```
Comment (bruth) [inline README.md]: Add a newline above for clarity.

File: README.md [hunk 13]
```diff
@@ -6,14 +6,140 @@ A validator for CSV files adhering to the [Data Models](https://github.com/chop-
...
+```
+
+If the header does not match the expected set of fields, the expected and actual number of fields as well as any unknown and/or missing fields found in the header are printed:
+
+```
+$ data-models-validator -model pedsnet -version 2.0.0 original_visit_occurrence.csv:visit_occurrence
+Validating against model 'pedsnet/2.0.0'
+* Evaluating 'visit_occurrence' table in 'original_visit_occurrence.csv'...
+* Problem reading CSV header: line 0: [code: 201] Header does not contain the correct set of fields: map[expectedLength:12 actualLength:13 unknownFields:[visit_occurrence_source_id] missingFields:[]]
```
Comment (bruth) [inline README.md]: The `map[...]` structure is leaking through here.

Comment (aaron0browne) [inline README.md]: I'll work on a fancier print statement [here](https://github.com/chop-dbhi/data-models-validator/pull/7/files#diff-dd25e0e0a594c3f3592c1c717039b85eR134).

**Comment by aaron0browne** [comment]
Changes re comments incorporated.

**Comment by aaron0browne** [comment]
Context formatting moved.

**Comment by aaron0browne** [comment]
Updated context formatting code.

File: README.md [hunk 14]
```diff
@@ -6,14 +6,122 @@ A validator for CSV files adhering to the [Data Models](https://github.com/chop-
...
+$ data-models-validator -model pedsnet -version 2.0.0 PEDSNET_DRUG_EXPOSURE.csv:drug_exposure
+```
+
+If the header does not match the expected set of fields, the expected and actual number of fields as well as any unknown and/or missing fields found in the header are printed:
+
+```
+$ data-models-validator -model pedsnet -version 2.0.0 original_visit_occurrence.csv:visit_occurrence
+Validating against model 'pedsnet/2.0.0'
+* Evaluating 'visit_occurrence' table in 'original_visit_occurrence.csv'...
+* Problem reading CSV header: line 0: [code: 201] Header does not contain the correct set of fields: (expectedLength:12, actualLength:13, unknownFields:[visit_occurrence_source_id], missingFields:[])
```
Comment (bruth) [inline README.md]: The empty `missingFields` array is a side effect of the `Context.String` method. We can add a function to detect for a [zero or nil value](https://golang.org/ref/spec#The_zero_value).

```go
func isZeroValue(x interface{}) bool {
    return x == nil || x == reflect.Zero(reflect.TypeOf(x)).Interface()
}
```

In the `Context.String` method, add a conditional to add a token if `isZeroValue` is false.

Comment (aaron0browne) [inline README.md]: I think as a user its actually nice to see that there are no missing fields in the header. Would you be OK with leaving that it?

Comment (bruth) [inline README.md]: Users don't know *a priori* that they will be provided that information, so not seeing that information translates that nothing else is wrong. If you really want to keep it, ensure that in other error contexts, empty values are not be added.

**Comment by aaron0browne** [comment]
OK, we won't print the empty values. You convinced me.

**Comment by aaron0browne** [comment]
Squashed. Should I update the version to 1.0.0 beta in this branch? Then I would think we would update to 1.0.0 final once the commits are in master and tag and release on that version update commit.

**Comment by bruth** [comment]
Yes.

**Comment by aaron0browne** [comment]
What is the intent of [`progReleaseNum`](https://github.com/chop-dbhi/data-models-validator/blob/0337c711edf3fa6da0a749a988b21b3ece2712b5/version.go#L11)? What should I set it to?

**Comment by bruth** [comment]
There really isn't. You can remove it to reduce confusion and we also have the `progBuild` string.

**Comment by aaron0browne** [comment]
OK, I think this is ready for merge.

**Comment by bruth** [comment]
Well done!
